src/requiredBy.js

```diff
@@ -0,0 +1,12 @@
+export default function requiredBy(requiredByPropName, propType, defaultValue = null) {
+  return function validator(props, propName, componentName, ...rest) {
+    if (props[requiredByPropName]) {
+      if (props[propName] === defaultValue || typeof props[propName] === 'undefined') {
+        return new TypeError(
+          `${componentName}: when ${requiredByPropName} is true, prop “${propName}” must be present.`,
+        );
+      }
+    }
+    return propType(props, propName, componentName, ...rest);
+  };
```

Comment:
I think that it doesn't need to block this PR :-)

but I may want to add in an `.isRequired` afterwards just to maintain api consistency. we can chat more offline about it perhaps.